src/Cms/FileActions.php

```diff
@@ -151,35 +151,61 @@ public function changeTemplate(string|null $template): static
 	/**
 	 * Commits a file action, by following these steps
 	 *
-	 * 1. checks the action rules
-	 * 2. sends the before hook
+	 * 1. applies the `before` hook
+	 * 2. checks the action rules
 	 * 3. commits the store action
-	 * 4. sends the after hook
+	 * 4. applies the `after` hook
 	 * 5. returns the result
 	 */
 	protected function commit(
 		string $action,
 		array $arguments,
 		Closure $callback
 	): mixed {
-		$old            = $this->hardcopy();
-		$kirby          = $this->kirby();
+		$kirby = $this->kirby();
+
+		// store copy of the model to be passed
+		// to the `after` hook for comparison
+		$old = $this->hardcopy();
+
+		// run `before` hook and pass all arguments;
+		// the very first argument (which should be the model)
+		// is modified by the return value from the hook (if any returned)
+		$appliedTo = array_key_first($arguments);
+		$arguments[$appliedTo] = $kirby->apply(
+			'file.' . $action . ':before',
+			$arguments,
+			$appliedTo
+		);
+
+		// get just the values of argument, so that spreading them
+		// as arguments in a method call gets only assigned by order
+		// not by key name
 		$argumentValues = array_values($arguments);
 
+		// check file rules
 		$this->rules()->$action(...$argumentValues);
```

Comment:
`Question:` Should the rules run twice? Before and after the `before` hook?

Reply:
Hm, very good question. I think the hook should really only run if the model passes examination. That's why I would say yes. But I'm not sure if running the rules twice could get us into trouble somehow. On the other hand, the hook really has the potential of changing the model in a way that would needs to be checked and maybe rejected again. I think I stick to yes: before and after the hook is probably needed.